In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# Load the dataset
df_online_retail_data = pd.read_csv('online_retail.csv')
df_online_retail_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

                InvoiceDate  UnitPrice  CustomerID         Country  
0       2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1       2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2       2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3       2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4       2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                     ...        ...         ...             ...  
541904  2011-12-09 12:50:00       0.85     12680.0          France  
541905  2011-12-09 12:50:00       2.10     12680.0          France  
541906  2011-12-09 12:50:00       4.15     12680.0          France  
541907  2011-12-09 12:50:00       4.15     12680.0          France  
541908  2011-12-09 12:50:00       4.95     12680.0          France  

[541909 rows x 8 columns]

Completar los datos nulos de la Serie Description

In [3]:
# Crear una copia del DataFrame
df_online_retail_data_filldesc = df_online_retail_data.copy()

In [4]:
# Rellenar los valores nulos en la columna 'Description' con valores de la columna 'StockCode'
df_online_retail_data_filldesc["Description"] = df_online_retail_data_filldesc.groupby("StockCode")["Description"].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else "Descripción desconocida"))


In [5]:
df_online_retail_data_filldesc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  541909 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [6]:
# Verificar 10 valores nulos del df original
df_online_retail_data[df_online_retail_data["Description"].isnull()].head(10)


InvoiceNo StockCode Description  Quantity          InvoiceDate  \
622     536414     22139         NaN        56  2010-12-01 11:52:00   
1970    536545     21134         NaN         1  2010-12-01 14:32:00   
1971    536546     22145         NaN         1  2010-12-01 14:33:00   
1972    536547     37509         NaN         1  2010-12-01 14:33:00   
1987    536549    85226A         NaN         1  2010-12-01 14:34:00   
1988    536550     85044         NaN         1  2010-12-01 14:34:00   
2024    536552     20950         NaN         1  2010-12-01 14:34:00   
2025    536553     37461         NaN         3  2010-12-01 14:35:00   
2026    536554     84670         NaN        23  2010-12-01 14:35:00   
2406    536589     21777         NaN       -10  2010-12-01 16:50:00   

      UnitPrice  CustomerID         Country  
622         0.0         NaN  United Kingdom  
1970        0.0         NaN  United Kingdom  
1971        0.0         NaN  United Kingdom  
1972        0.0         NaN  United Kingdom  
1987        0.0         NaN  United Kingdom  
1988        0.0         NaN  United Kingdom  
2024        0.0         NaN  United Kingdom  
2025        0.0         NaN  United Kingdom  
2026        0.0         NaN  United Kingdom  
2406        0.0         NaN  United Kingdom

In [7]:
# Verificar 10 valores nulos del df modificado con iloc
print(df_online_retail_data_filldesc.iloc[1970:1988])

     InvoiceNo StockCode                          Description  Quantity  \
1970    536545     21134              Descripción desconocida         1   
1971    536546     22145      CHRISTMAS CRAFT HEART STOCKING          1   
1972    536547     37509           NEW ENGLAND MUG W GIFT BOX         1   
1973   C536548     22244           3 HOOK HANGER MAGIC GARDEN        -4   
1974   C536548     22242        5 HOOK HANGER MAGIC TOADSTOOL        -5   
1975   C536548     20914  SET/5 RED RETROSPOT LID GLASS BOWLS        -1   
1976   C536548     22892    SET OF SALT AND PEPPER TOADSTOOLS        -7   
1977   C536548     22654                   DELUXE SEWING KIT         -1   
1978   C536548     22767          TRIPLE PHOTO FRAME CORNICE         -2   
1979   C536548     22333    RETROSPOT PARTY BAG + STICKER SET        -1   
1980   C536548     22245         HOOK, 1 HANGER ,MAGIC GARDEN        -2   
1981   C536548     22077               6 RIBBONS RUSTIC CHARM        -6   
1982   C536548     22631 

In [8]:
print(df_online_retail_data_filldesc.iloc[2024:2027])

     InvoiceNo StockCode              Description  Quantity  \
2024    536552     20950  Descripción desconocida         1   
2025    536553     37461         FUNKY MONKEY MUG         3   
2026    536554     84670  Descripción desconocida        23   

              InvoiceDate  UnitPrice  CustomerID         Country  
2024  2010-12-01 14:34:00        0.0         NaN  United Kingdom  
2025  2010-12-01 14:35:00        0.0         NaN  United Kingdom  
2026  2010-12-01 14:35:00        0.0         NaN  United Kingdom  


In [9]:
# Crear un DataFrame con las filas donde la descripción es "Descripción desconocida"
df_descripcion_desconocida = df_online_retail_data_filldesc[df_online_retail_data_filldesc["Description"] == "Descripción desconocida"]

In [10]:
df_descripcion_desconocida.shape

(112, 8)

Nota: Una vez rellenado los datos nulos, se tienen ahora datos con "Descripcion Desconocida".  Son 112 datos que por su tamano se considera que no son influyentes.  En tal sentido, Se eliminaran de DF.

In [11]:
# Eliminar los datos con "Descripción desconocida" del DataFrame df_online_retail_data_filldesc
df_online_retail_data_filldesc.drop(df_online_retail_data_filldesc[df_online_retail_data_filldesc['Description'] == 'Descripción desconocida'].index, inplace=True)

In [12]:
# Filtras datos de una columna 
df_online_retail_data_filldesc[df_online_retail_data_filldesc['Description'] == 'Descripción desconocida']

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

Trabajar sobre el Dataframe df_online_retail_data_filldesc para trabajar los datos nulos de la columna CustomerId

In [13]:
df_online_retail_data_filldesc.isnull().sum()

InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     134968
Country             0
dtype: int64

In [14]:
df_online_retail_data_filldesc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 541797 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541797 non-null  object 
 1   StockCode    541797 non-null  object 
 2   Description  541797 non-null  object 
 3   Quantity     541797 non-null  int64  
 4   InvoiceDate  541797 non-null  object 
 5   UnitPrice    541797 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541797 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 37.2+ MB


In [15]:
print("Unique values in 'CustomerID':", df_online_retail_data['CustomerID'].nunique())
print("null values in 'CustomerID':", df_online_retail_data['CustomerID'].isnull().sum())

Unique values in 'CustomerID': 4372
null values in 'CustomerID': 135080


Completar los valores nulos en 'CustomerID' relacionandolo con el InvoivceNo

In [16]:
# Duplicarl el DataFrame df_online_retail_data_filldesc como un DataFrame temporal
df_online_retail_data_trans = df_online_retail_data_filldesc.copy()
df_online_retail_data_trans.info()

<class 'pandas.core.frame.DataFrame'>
Index: 541797 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541797 non-null  object 
 1   StockCode    541797 non-null  object 
 2   Description  541797 non-null  object 
 3   Quantity     541797 non-null  int64  
 4   InvoiceDate  541797 non-null  object 
 5   UnitPrice    541797 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541797 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 37.2+ MB


In [17]:
# Rellenar los valores nulos en la columna 'CustomerID' con valores de la columna 'InvoiceNo'
df_online_retail_data_trans["CustomerID"] = df_online_retail_data_trans.groupby("InvoiceNo")["CustomerID"].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else 12345))


In [18]:
df_online_retail_data_trans.info()

<class 'pandas.core.frame.DataFrame'>
Index: 541797 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541797 non-null  object 
 1   StockCode    541797 non-null  object 
 2   Description  541797 non-null  object 
 3   Quantity     541797 non-null  int64  
 4   InvoiceDate  541797 non-null  object 
 5   UnitPrice    541797 non-null  float64
 6   CustomerID   541797 non-null  float64
 7   Country      541797 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 37.2+ MB


In [19]:
df_online_retail_data_trans[df_online_retail_data_trans["CustomerID"] == "12345"].head(10)

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

Convertir el DataFrame ya transformado en un archivo .csv para hacer los análisis del proyecto

In [20]:
df_online_retail_data_trans.to_csv('online_retail_transformed.csv', index=False)

NOTA: Aqui debo reviar... me falta eliminar los UnitPrice Negativos y validar los UnitPrice cero.  Esta exportacion no se va a usar pero por factor Q, no la borro

Limpieza de los Datos UnitPrice

In [21]:
# Eliminar los datos con "UnitPrice negativos" del DataFrame df_online_retail_data_trans
df_online_retail_data_trans.drop(df_online_retail_data_trans[df_online_retail_data_trans['UnitPrice'] <0].index, inplace=True)

In [22]:
df_online_retail_data_trans.info()

<class 'pandas.core.frame.DataFrame'>
Index: 541795 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541795 non-null  object 
 1   StockCode    541795 non-null  object 
 2   Description  541795 non-null  object 
 3   Quantity     541795 non-null  int64  
 4   InvoiceDate  541795 non-null  object 
 5   UnitPrice    541795 non-null  float64
 6   CustomerID   541795 non-null  float64
 7   Country      541795 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 37.2+ MB


In [23]:
df_online_retail_data_trans.describe()

Quantity      UnitPrice     CustomerID
count  541795.00000  541795.000000  541795.000000
mean        9.55595       4.652919   14554.639923
std       218.10383      94.405179    1955.687994
min    -80995.00000       0.000000   12345.000000
25%         1.00000       1.250000   12356.000000
50%         3.00000       2.080000   14382.000000
75%        10.00000       4.130000   16255.000000
max     80995.00000   38970.000000   18287.000000

Se aprecia una incongruencia para el valor max en la columna UnitPrice.  Es un dato Outlier.  

In [24]:
df_online_retail_data_trans.sort_values(by='UnitPrice', ascending=False).head(30)

InvoiceNo  StockCode      Description  Quantity          InvoiceDate  \
222681   C556445          M           Manual        -1  2011-06-10 15:31:00   
524602   C580605  AMAZONFEE       AMAZON FEE        -1  2011-12-05 11:36:00   
43702    C540117  AMAZONFEE       AMAZON FEE        -1  2011-01-05 09:55:00   
43703    C540118  AMAZONFEE       AMAZON FEE        -1  2011-01-05 09:57:00   
16356    C537651  AMAZONFEE       AMAZON FEE        -1  2010-12-07 15:49:00   
15016    C537630  AMAZONFEE       AMAZON FEE        -1  2010-12-07 15:04:00   
15017     537632  AMAZONFEE       AMAZON FEE         1  2010-12-07 15:08:00   
16232    C537644  AMAZONFEE       AMAZON FEE        -1  2010-12-07 15:34:00   
524601   C580604  AMAZONFEE       AMAZON FEE        -1  2011-12-05 11:35:00   
299982   A563185          B  Adjust bad debt         1  2011-08-12 14:50:00   
446533   C574902  AMAZONFEE       AMAZON FEE        -1  2011-11-07 15:21:00   
173382    551697       POST          POSTAGE         1  2011-05-03 13:46:00   
173277   C551685       POST          POSTAGE        -1  2011-05-03 12:51:00   
342635   C566899  AMAZONFEE       AMAZON FEE        -1  2011-09-15 13:53:00   
191386   C553355  AMAZONFEE       AMAZON FEE        -1  2011-05-16 13:58:00   
173391   C551699          M           Manual        -1  2011-05-03 14:12:00   
287150   C562086  AMAZONFEE       AMAZON FEE        -1  2011-08-02 12:27:00   
16357    C537652  AMAZONFEE       AMAZON FEE        -1  2010-12-07 15:51:00   
312246   C564341  AMAZONFEE       AMAZON FEE        -1  2011-08-24 14:53:00   
262414   C559917  AMAZONFEE       AMAZON FEE        -1  2011-07-13 15:21:00   
383495   C570025  AMAZONFEE       AMAZON FEE        -1  2011-10-07 10:29:00   
429248   C573549  AMAZONFEE       AMAZON FEE        -1  2011-10-31 13:23:00   
446434   C574897  AMAZONFEE       AMAZON FEE        -1  2011-11-07 15:03:00   
191385   C553354  AMAZONFEE       AMAZON FEE        -1  2011-05-16 13:54:00   
239250   C558036  AMAZONFEE       AMAZON FEE        -1  2011-06-24 12:31:00   
124741   C546987  AMAZONFEE       AMAZON FEE        -1  2011-03-18 12:56:00   
96844    C544587  AMAZONFEE       AMAZON FEE        -1  2011-02-21 15:07:00   
342611   C566889  AMAZONFEE       AMAZON FEE        -1  2011-09-15 13:50:00   
16313    C537647  AMAZONFEE       AMAZON FEE        -1  2010-12-07 15:41:00   
96845    C544589  AMAZONFEE       AMAZON FEE        -1  2011-02-21 15:11:00   

        UnitPrice  CustomerID         Country  
222681   38970.00     15098.0  United Kingdom  
524602   17836.46     12345.0  United Kingdom  
43702    16888.02     12345.0  United Kingdom  
43703    16453.71     12345.0  United Kingdom  
16356    13541.33     12345.0  United Kingdom  
15016    13541.33     12345.0  United Kingdom  
15017    13541.33     12345.0  United Kingdom  
16232    13474.79     12345.0  United Kingdom  
524601   11586.50     12345.0  United Kingdom  
299982   11062.06     12345.0  United Kingdom  
446533    8286.22     12345.0  United Kingdom  
173382    8142.75     16029.0  United Kingdom  
173277    8142.75     16029.0  United Kingdom  
342635    7427.97     12345.0  United Kingdom  
191386    7006.83     12345.0  United Kingdom  
173391    6930.00     16029.0  United Kingdom  
287150    6721.37     12345.0  United Kingdom  
16357     6706.71     12345.0  United Kingdom  
312246    6662.51     12345.0  United Kingdom  
262414    6497.47     12345.0  United Kingdom  
383495    5942.57     12345.0  United Kingdom  
429248    5942.57     12345.0  United Kingdom  
446434    5877.18     12345.0  United Kingdom  
191385    5876.40     12345.0  United Kingdom  
239250    5791.18     12345.0  United Kingdom  
124741    5693.05     12345.0  United Kingdom  
96844     5575.28     12345.0  United Kingdom  
342611    5522.14     12345.0  United Kingdom  
16313     5519.25     12345.0  United Kingdom  
96845     5258.77     12345.0  United Kingdom

- Se aprecia como un error el invoice C556445 y se procedera a eliminar este deto
- Los datos en descrition como AMAZON FEE no son ventas.  Por lo cual se eliminaran

In [25]:
df_online_retail_data_trans.drop(df_online_retail_data_trans[df_online_retail_data_trans['InvoiceNo'] == 'C556445'].index, inplace=True)

In [26]:
df_online_retail_data_trans.sort_values(by='UnitPrice', ascending=False).head(30)

InvoiceNo  StockCode      Description  Quantity          InvoiceDate  \
524602   C580605  AMAZONFEE       AMAZON FEE        -1  2011-12-05 11:36:00   
43702    C540117  AMAZONFEE       AMAZON FEE        -1  2011-01-05 09:55:00   
43703    C540118  AMAZONFEE       AMAZON FEE        -1  2011-01-05 09:57:00   
15017     537632  AMAZONFEE       AMAZON FEE         1  2010-12-07 15:08:00   
15016    C537630  AMAZONFEE       AMAZON FEE        -1  2010-12-07 15:04:00   
16356    C537651  AMAZONFEE       AMAZON FEE        -1  2010-12-07 15:49:00   
16232    C537644  AMAZONFEE       AMAZON FEE        -1  2010-12-07 15:34:00   
524601   C580604  AMAZONFEE       AMAZON FEE        -1  2011-12-05 11:35:00   
299982   A563185          B  Adjust bad debt         1  2011-08-12 14:50:00   
446533   C574902  AMAZONFEE       AMAZON FEE        -1  2011-11-07 15:21:00   
173382    551697       POST          POSTAGE         1  2011-05-03 13:46:00   
173277   C551685       POST          POSTAGE        -1  2011-05-03 12:51:00   
342635   C566899  AMAZONFEE       AMAZON FEE        -1  2011-09-15 13:53:00   
191386   C553355  AMAZONFEE       AMAZON FEE        -1  2011-05-16 13:58:00   
173391   C551699          M           Manual        -1  2011-05-03 14:12:00   
287150   C562086  AMAZONFEE       AMAZON FEE        -1  2011-08-02 12:27:00   
16357    C537652  AMAZONFEE       AMAZON FEE        -1  2010-12-07 15:51:00   
312246   C564341  AMAZONFEE       AMAZON FEE        -1  2011-08-24 14:53:00   
262414   C559917  AMAZONFEE       AMAZON FEE        -1  2011-07-13 15:21:00   
429248   C573549  AMAZONFEE       AMAZON FEE        -1  2011-10-31 13:23:00   
383495   C570025  AMAZONFEE       AMAZON FEE        -1  2011-10-07 10:29:00   
446434   C574897  AMAZONFEE       AMAZON FEE        -1  2011-11-07 15:03:00   
191385   C553354  AMAZONFEE       AMAZON FEE        -1  2011-05-16 13:54:00   
239250   C558036  AMAZONFEE       AMAZON FEE        -1  2011-06-24 12:31:00   
124741   C546987  AMAZONFEE       AMAZON FEE        -1  2011-03-18 12:56:00   
96844    C544587  AMAZONFEE       AMAZON FEE        -1  2011-02-21 15:07:00   
342611   C566889  AMAZONFEE       AMAZON FEE        -1  2011-09-15 13:50:00   
16313    C537647  AMAZONFEE       AMAZON FEE        -1  2010-12-07 15:41:00   
96845    C544589  AMAZONFEE       AMAZON FEE        -1  2011-02-21 15:11:00   
124787   C546989  AMAZONFEE       AMAZON FEE        -1  2011-03-18 12:59:00   

        UnitPrice  CustomerID         Country  
524602   17836.46     12345.0  United Kingdom  
43702    16888.02     12345.0  United Kingdom  
43703    16453.71     12345.0  United Kingdom  
15017    13541.33     12345.0  United Kingdom  
15016    13541.33     12345.0  United Kingdom  
16356    13541.33     12345.0  United Kingdom  
16232    13474.79     12345.0  United Kingdom  
524601   11586.50     12345.0  United Kingdom  
299982   11062.06     12345.0  United Kingdom  
446533    8286.22     12345.0  United Kingdom  
173382    8142.75     16029.0  United Kingdom  
173277    8142.75     16029.0  United Kingdom  
342635    7427.97     12345.0  United Kingdom  
191386    7006.83     12345.0  United Kingdom  
173391    6930.00     16029.0  United Kingdom  
287150    6721.37     12345.0  United Kingdom  
16357     6706.71     12345.0  United Kingdom  
312246    6662.51     12345.0  United Kingdom  
262414    6497.47     12345.0  United Kingdom  
429248    5942.57     12345.0  United Kingdom  
383495    5942.57     12345.0  United Kingdom  
446434    5877.18     12345.0  United Kingdom  
191385    5876.40     12345.0  United Kingdom  
239250    5791.18     12345.0  United Kingdom  
124741    5693.05     12345.0  United Kingdom  
96844     5575.28     12345.0  United Kingdom  
342611    5522.14     12345.0  United Kingdom  
16313     5519.25     12345.0  United Kingdom  
96845     5258.77     12345.0  United Kingdom  
124787    5225.03     12345.0  United Kingdom

In [27]:
df_online_retail_data_trans.drop(df_online_retail_data_trans[df_online_retail_data_trans['StockCode'] == 'AMAZONFEE'].index, inplace=True)

In [28]:
df_online_retail_data_trans.sort_values(by='UnitPrice', ascending=False).head(30)

InvoiceNo StockCode      Description  Quantity          InvoiceDate  \
299982   A563185         B  Adjust bad debt         1  2011-08-12 14:50:00   
173277   C551685      POST          POSTAGE        -1  2011-05-03 12:51:00   
173382    551697      POST          POSTAGE         1  2011-05-03 13:46:00   
173391   C551699         M           Manual        -1  2011-05-03 14:12:00   
297723    562955       DOT   DOTCOM POSTAGE         1  2011-08-11 10:14:00   
268028    560373         M           Manual         1  2011-07-18 12:30:00   
268027   C560372         M           Manual        -1  2011-07-18 12:26:00   
422376    573080         M           Manual         1  2011-10-27 14:20:00   
422375   C573079         M           Manual        -2  2011-10-27 14:15:00   
422351    573077         M           Manual         1  2011-10-27 14:13:00   
406406    571751         M           Manual         1  2011-10-19 11:18:00   
406404   C571750         M           Manual        -1  2011-10-19 11:16:00   
374542    569382         M           Manual         1  2011-10-03 16:44:00   
293842   C562647         M           Manual        -1  2011-08-08 12:56:00   
271151   C560647         M           Manual        -1  2011-07-20 11:31:00   
347947   C567352         M           Manual        -1  2011-09-19 16:13:00   
347948    567353         M           Manual         1  2011-09-19 16:14:00   
119631   C546557         M           Manual        -1  2011-03-15 09:44:00   
119632    546558         M           Manual         1  2011-03-15 09:50:00   
297438    562946         M           Manual         1  2011-08-11 09:38:00   
144826    548813         M           Manual         1  2011-04-04 13:03:00   
144831   C548830         M           Manual        -1  2011-04-04 13:08:00   
493020    578149       DOT   DOTCOM POSTAGE         1  2011-11-23 11:11:00   
524892    580610       DOT   DOTCOM POSTAGE         1  2011-12-05 11:48:00   
406407    571751         M           Manual         1  2011-10-19 11:18:00   
406405   C571750         M           Manual        -1  2011-10-19 11:16:00   
525134    580612       DOT   DOTCOM POSTAGE         1  2011-12-05 11:58:00   
144829    548820         M           Manual         1  2011-04-04 13:04:00   
144834   C548834         M           Manual        -1  2011-04-04 13:09:00   
343000    566927         M           Manual         1  2011-09-15 15:20:00   

        UnitPrice  CustomerID         Country  
299982   11062.06     12345.0  United Kingdom  
173277    8142.75     16029.0  United Kingdom  
173382    8142.75     16029.0  United Kingdom  
173391    6930.00     16029.0  United Kingdom  
297723    4505.17     12345.0  United Kingdom  
268028    4287.63     12345.0  United Kingdom  
268027    4287.63     17448.0  United Kingdom  
422376    4161.06     12536.0          France  
422375    4161.06     12536.0          France  
422351    4161.06     12536.0          France  
406406    3949.32     12744.0       Singapore  
406404    3949.32     12744.0       Singapore  
374542    3155.95     15502.0  United Kingdom  
293842    3155.95     15502.0  United Kingdom  
271151    3060.60     18102.0  United Kingdom  
347947    2653.95     12345.0       Hong Kong  
347948    2653.95     12345.0       Hong Kong  
119631    2583.76     12345.0       Hong Kong  
119632    2583.76     12345.0       Hong Kong  
297438    2500.00     15581.0  United Kingdom  
144826    2382.92     12744.0       Singapore  
144831    2382.92     12744.0       Singapore  
493020    2275.54     12345.0  United Kingdom  
524892    2196.67     12345.0  United Kingdom  
406407    2118.74     12744.0       Singapore  
406405    2118.74     12744.0       Singapore  
525134    2114.00     12345.0  United Kingdom  
144829    2053.07     12744.0       Singapore  
144834    2053.07     12744.0       Singapore  
343000    2033.10     17846.0  United Kingdom

In [29]:
df_online_retail_data_trans.describe()

Quantity      UnitPrice     CustomerID
count  541760.000000  541760.000000  541760.000000
mean        9.556625       4.121595   14554.777593
std       218.110859      38.043811    1955.672682
min    -80995.000000       0.000000   12345.000000
25%         1.000000       1.250000   12356.000000
50%         3.000000       2.080000   14382.000000
75%        10.000000       4.130000   16255.000000
max     80995.000000   11062.060000   18287.000000

Para tener los datos con una media mas precisa, y revisando con el cliente, se determina que el dato InvoiceNo = A563185 sera eliminado. 

In [30]:
df_online_retail_data_trans.drop(df_online_retail_data_trans[df_online_retail_data_trans['InvoiceNo'] == 'A563185'].index, inplace=True)

In [31]:
df_online_retail_data_trans.sort_values(by='UnitPrice', ascending=False).head(30)

InvoiceNo StockCode     Description  Quantity          InvoiceDate  \
173277   C551685      POST         POSTAGE        -1  2011-05-03 12:51:00   
173382    551697      POST         POSTAGE         1  2011-05-03 13:46:00   
173391   C551699         M          Manual        -1  2011-05-03 14:12:00   
297723    562955       DOT  DOTCOM POSTAGE         1  2011-08-11 10:14:00   
268027   C560372         M          Manual        -1  2011-07-18 12:26:00   
268028    560373         M          Manual         1  2011-07-18 12:30:00   
422351    573077         M          Manual         1  2011-10-27 14:13:00   
422375   C573079         M          Manual        -2  2011-10-27 14:15:00   
422376    573080         M          Manual         1  2011-10-27 14:20:00   
406404   C571750         M          Manual        -1  2011-10-19 11:16:00   
406406    571751         M          Manual         1  2011-10-19 11:18:00   
293842   C562647         M          Manual        -1  2011-08-08 12:56:00   
374542    569382         M          Manual         1  2011-10-03 16:44:00   
271151   C560647         M          Manual        -1  2011-07-20 11:31:00   
347947   C567352         M          Manual        -1  2011-09-19 16:13:00   
347948    567353         M          Manual         1  2011-09-19 16:14:00   
119632    546558         M          Manual         1  2011-03-15 09:50:00   
119631   C546557         M          Manual        -1  2011-03-15 09:44:00   
297438    562946         M          Manual         1  2011-08-11 09:38:00   
144831   C548830         M          Manual        -1  2011-04-04 13:08:00   
144826    548813         M          Manual         1  2011-04-04 13:03:00   
493020    578149       DOT  DOTCOM POSTAGE         1  2011-11-23 11:11:00   
524892    580610       DOT  DOTCOM POSTAGE         1  2011-12-05 11:48:00   
406407    571751         M          Manual         1  2011-10-19 11:18:00   
406405   C571750         M          Manual        -1  2011-10-19 11:16:00   
525134    580612       DOT  DOTCOM POSTAGE         1  2011-12-05 11:58:00   
144834   C548834         M          Manual        -1  2011-04-04 13:09:00   
144829    548820         M          Manual         1  2011-04-04 13:04:00   
343000    566927         M          Manual         1  2011-09-15 15:20:00   
502058    578833       DOT  DOTCOM POSTAGE         1  2011-11-25 15:23:00   

        UnitPrice  CustomerID         Country  
173277    8142.75     16029.0  United Kingdom  
173382    8142.75     16029.0  United Kingdom  
173391    6930.00     16029.0  United Kingdom  
297723    4505.17     12345.0  United Kingdom  
268027    4287.63     17448.0  United Kingdom  
268028    4287.63     12345.0  United Kingdom  
422351    4161.06     12536.0          France  
422375    4161.06     12536.0          France  
422376    4161.06     12536.0          France  
406404    3949.32     12744.0       Singapore  
406406    3949.32     12744.0       Singapore  
293842    3155.95     15502.0  United Kingdom  
374542    3155.95     15502.0  United Kingdom  
271151    3060.60     18102.0  United Kingdom  
347947    2653.95     12345.0       Hong Kong  
347948    2653.95     12345.0       Hong Kong  
119632    2583.76     12345.0       Hong Kong  
119631    2583.76     12345.0       Hong Kong  
297438    2500.00     15581.0  United Kingdom  
144831    2382.92     12744.0       Singapore  
144826    2382.92     12744.0       Singapore  
493020    2275.54     12345.0  United Kingdom  
524892    2196.67     12345.0  United Kingdom  
406407    2118.74     12744.0       Singapore  
406405    2118.74     12744.0       Singapore  
525134    2114.00     12345.0  United Kingdom  
144834    2053.07     12744.0       Singapore  
144829    2053.07     12744.0       Singapore  
343000    2033.10     17846.0  United Kingdom  
502058    2028.25     12345.0  United Kingdom

In [32]:
df_online_retail_data_trans.describe()

Quantity      UnitPrice     CustomerID
count  541759.00000  541759.000000  541759.000000
mean        9.55664       4.101184   14554.781672
std       218.11106      34.951793    1955.672182
min    -80995.00000       0.000000   12345.000000
25%         1.00000       1.250000   12356.000000
50%         3.00000       2.080000   14382.000000
75%        10.00000       4.130000   16255.000000
max     80995.00000    8142.750000   18287.000000

Concluion: Se trabajara con este DataFrame limpio / transformado.  EN acuerdo con el cliente, los datos POSTAGE y DOTCOM POSTAGE se mantendran.  

In [33]:
# Copiar el DataFrame para usarlo en el analisis del proyecto
df_online_retail_cleaned = df_online_retail_data_trans.copy()
df_online_retail_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 541759 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541759 non-null  object 
 1   StockCode    541759 non-null  object 
 2   Description  541759 non-null  object 
 3   Quantity     541759 non-null  int64  
 4   InvoiceDate  541759 non-null  object 
 5   UnitPrice    541759 non-null  float64
 6   CustomerID   541759 non-null  float64
 7   Country      541759 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 37.2+ MB


In [34]:
df_online_retail_cleaned.describe()

Quantity      UnitPrice     CustomerID
count  541759.00000  541759.000000  541759.000000
mean        9.55664       4.101184   14554.781672
std       218.11106      34.951793    1955.672182
min    -80995.00000       0.000000   12345.000000
25%         1.00000       1.250000   12356.000000
50%         3.00000       2.080000   14382.000000
75%        10.00000       4.130000   16255.000000
max     80995.00000    8142.750000   18287.000000

In [35]:
df_online_retail_cleaned.to_csv('online_retail_cleaned.csv', index=False)